### PR TITLE
euler's totient function in python

### PR DESCRIPTION
**Pull request by ag278**
### Have you read the [Contributing Guidelines on Pull Requests](https://github.com/TesseractCoding/NeoAlgo/blob/master/CONTRIBUTING.md#pull-requests)?

Yes

### Description

I have created Euler's_Totient_Function.cpp file in python folder and I have updated readme file.

### Checklist

- [x] I've read the contribution guidelines.
- [x] I've checked the issue list before deciding what to submit.
- [x] I've edited the `README.md` and link to my code.

## Related Issues or Pull Requests

#445

### CODE REVIEW

File: Python/README.md [hunk 1]
```diff
@@ -64,6 +64,7 @@
 * [Sieve of Eratosthenes](math/prime_number_using_sieve_of_Eratosthenes.py)
 * [Fibonacci number](math/Fibonacci.py)
 * [Strassen's Matrix Multiplication](math/Strassen's_Matrix_Mul.py)
+* [Euler’s Totient Function](euler_totient_function.py)
```
Comment (atarax665) [inline Python/README.md]: It should be **(math/euler_totient_function.py)**

**Review by atarax665 — CHANGES_REQUESTED** [review]

File: Python/math/euler_totient_function.py [hunk 2]
```diff
@@ -0,0 +1,21 @@
+# python program for Euler's Totient Function 
+import math
+def euler_totient(q):
+    ans=q   #defining the answer initially to n
```
Comment (atarax665) [inline Python/math/euler_totient_function.py]: The function give wrong answer for various sample cases. ex- 4, 6, 9 etc. Please check the algorithm used once again.

File: Python/math/euler_totient_function.py [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+     ans-=ans/q            
+     
+    print(int(ans))
+    
+print("Enter number of elements : ")
+n = int(input())           # taking input from the user
+if(n<0):                   
+    print("inavlid input")
+else:
+    euler_totient(n)
```
Comment (atarax665) [inline Python/math/euler_totient_function.py]: Add sample input/output in the end.

File: Python/math/euler_totient_function.py [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+import math
+def euler_totient(q):
+    ans=q   #defining the answer initially to n
+    for i in range (2,int(math.sqrt(q+1))):
+      if (q%i == 0):    #checking if i is a factor of q
+        while (q%i == 0): 
+          q/=i             # if the condition is satisified we are updating the value of n
+        ans-= ans/i        # just as we do in sieve of eratosthenes algorithm
+                           # we are decreasing values for the multiples of primes
+    if (q>1):              #this case ensures that if n have a prime factor greater than sqrt(n) then we are counting thaat case
```
Comment (atarax665) [inline Python/math/euler_totient_function.py]: Refer to [PEP 8](https://www.python.org/dev/peps/pep-0008/) to improve the readability of your code.

**Review by atarax665 — CHANGES_REQUESTED** [review]

**Comment by ag278** [comment]
Sir, actually I have noticed that there is a problem in the statement of the issue. I just check the name of the topic and since many of the people are commenting at that time i only notice the topic name and not the statement . Actually F(n) denotes the number of coprimes of n from 1 to n. Example F(4) denotes number of coprimes of 4 from 1 to 4, so F(4)=2 as only 1 & 3 are coprimes to 4. Actually I have written the code for  euler's totient function and not for the number of coprime number between 1 and n( as mentiones in the problem statement). So what should I do now?

**Comment by atarax665** [comment]
> Sir, actually I have noticed that there is a problem in the statement of the issue. I just check the name of the topic and since many of the people are commenting at that time i only notice the topic name and not the statement . Actually F(n) denotes the number of coprimes of n from 1 to n. Example F(4) denotes number of coprimes of 4 from 1 to 4, so F(4)=2 as only 1 & 3 are coprimes to 4. Actually I have written the code for euler's totient function and not for the number of coprime number between 1 and n( as mentiones in the problem statement). So what should I do now?

As this PR is to be merged under #445 I will suggest you to write a new function to count the number of coprime of n between 1 to n. You can refer to [euler's totient function](https://www.geeksforgeeks.org/eulers-totient-function/) .

**Comment by ag278** [comment]
@atarax665  i have made the changes in the file. Now the function euler_totient (n) calculates the no. of coprimes between 1 and n (inclusive of 1 and n).Please review it and if it is ok please let me know
Hakincode2020

**Comment by atarax665** [comment]
> @atarax665 i have made the changes in the file. Now the function euler_totient (n) calculates the no. of coprimes between 1 and n (inclusive of 1 and n).Please review it and if it is ok please let me know
> Hakincode2020

You had to calculate the no. of coprimes of n between 1 and n (both exclusive). Your previous code was giving wrong answer for various inputs.
euler
 For input = 6 the program should return 2 as the only gcd(1, 6) =1 and gcd(5, 6)=1.

File: Python/math/euler_totient_function.py [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+            for j in range(i,n+1,i):
+                arr[j]-=(arr[j]//i)  
+                
+    for i in range(2,n+1):
+        arr[i]+=arr[i-1]               #for calculating no of coprime between (1,n) we are using the concept of dyanmic programming 
+        
+    return arr[n]  
+    
+    
+print("Enter number of elements : ")
```
Comment (atarax665) [inline Python/math/euler_totient_function.py]: Change print statement to print('Enter a number : ')

**Review by atarax665 — CHANGES_REQUESTED** [review]

**Comment by ag278** [comment]
@atarax665 please verify it once.

File: Python/math/euler_totient_function.py [hunk 6]
```diff
@@ -0,0 +1,26 @@
+# python program for Euler's Totient Function 
+def euler_totient(q):
```
Comment (atarax665) [inline Python/math/euler_totient_function.py]: Add docstring to make the code look well documented and add space before and after operators.

**Review by atarax665 — CHANGES_REQUESTED** [review]

**Comment by ag278** [comment]
@atarax665 i have made the required changes. Please verify it once
hakincode 2020

File: Python/math/euler_totient_function.py [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+if(n < 0):                   
+    print("inavlid input")
+else:
+    print(euler_totient(n))
+    
+'''sample input
+euler_totient(4) will be 2 as coprime will be (1,4),(3,4)
+euler_totient(6) will be 2 as coprime will be (1,6),(5,6)
+euler_totient(9) will be 6 as coprime will be (1,9),(2,9),(4,9),(5,9),(7,9),(8,9)
+here we are considering (2,3) and (3,2) as same pair
```
Comment (atarax665) [inline Python/math/euler_totient_function.py:29]: Remove this line. Its redundant now. Everything else looks fine. :)

**Review by atarax665 — APPROVED** [review]

**Review by SKAUL05 — CHANGES_REQUESTED** [review]
Use meaningful variable names instead of alphabets

**Comment by ag278** [comment]
@SKAUL05  i have made the required changes. Please verify it once

File: Python/math/euler_totient_function.py [hunk 8]
```diff
@@ -0,0 +1,30 @@
+''' python program for Euler's Totient Function '''
+def euler_totient(num):
+    ''' Takes in a variable q and return the number of coprimes of n '''
+    Arr=[]
```
Comment (SKAUL05) [inline Python/math/euler_totient_function.py]: Change to arr ( lower case )

**Review by SKAUL05 — CHANGES_REQUESTED** [review]

**Comment by ag278** [comment]
@SKAUL05 i have made the required changes. Please verify it once

**Review by SKAUL05 — APPROVED** [review]

**Review by iamrajiv — APPROVED** [review]

**Comment by iamrajiv** [comment]
@all-contributors please add @ag278 for code

**Pull request merged**